File: google/cloud/dataqna/v1alpha/google-cloud-dataqna-v1alpha-ruby/test/google/cloud/data_qn_a/v1alpha/question_service_client_test.rb
# Copyright 2020 Google LLC
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     https://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

require "minitest/autorun"
require "minitest/spec"

require "google/gax"

require "google/cloud/data_qn_a/v1alpha"
require "google/cloud/data_qn_a/v1alpha/question_service_client"
require "google/cloud/dataqna/v1alpha/question_service_services_pb"

class CustomTestError_v1alpha < StandardError; end

# Mock for the GRPC::ClientStub class.
class MockGrpcClientStub_v1alpha

  # @param expected_symbol [Symbol] the symbol of the grpc method to be mocked.
  # @param mock_method [Proc] The method that is being mocked.
  def initialize(expected_symbol, mock_method)
    @expected_symbol = expected_symbol
    @mock_method = mock_method
  end

  # This overrides the Object#method method to return the mocked method when the mocked method
  # is being requested. For methods that aren't being tested, this method returns a proc that
  # will raise an error when called. This is to assure that only the mocked grpc method is being
  # called.
  #
  # @param symbol [Symbol] The symbol of the method being requested.
  # @return [Proc] The proc of the requested method. If the requested method is not being mocked
  #   the proc returned will raise when called.
  def method(symbol)
    return @mock_method if symbol == @expected_symbol

    # The requested method is not being tested, raise if it called.
    proc do
      raise "The method #{symbol} was unexpectedly called during the " \
        "test for #{@expected_symbol}."
    end
  end
end

class MockQuestionServiceCredentials_v1alpha < Google::Cloud::DataQnA::V1alpha::Credentials
  def initialize(method_name)
    @method_name = method_name
  end

  def updater_proc
    proc do
      raise "The method `#{@method_name}` was trying to make a grpc request. This should not " \
          "happen since the grpc layer is being mocked."
    end
  end
end

describe Google::Cloud::DataQnA::V1alpha::QuestionServiceClient do

  describe 'get_question' do
    custom_error = CustomTestError_v1alpha.new "Custom test error for Google::Cloud::DataQnA::V1alpha::QuestionServiceClient#get_question."

    it 'invokes get_question without error' do
      # Create request parameters
      formatted_name = Google::Cloud::DataQnA::V1alpha::QuestionServiceClient.question_path("[PROJECT]", "[LOCATION]", "[QUESTION]")

      # Create expected grpc response
      name_2 = "name2-1052831874"
      query = "query107944136"
      user_email = "userEmail1921668648"
      expected_response = {
        name: name_2,
        query: query,
        user_email: user_email
      }
      expected_response = Google::Gax::to_proto(expected_response, Google::Cloud::Dataqna::V1alpha::Question)

      # Mock Grpc layer
      mock_method = proc do |request|
        assert_instance_of(Google::Cloud::Dataqna::V1alpha::GetQuestionRequest, request)
        assert_equal(formatted_name, request.name)
        OpenStruct.new(execute: expected_response)
      end
      mock_stub = MockGrpcClientStub_v1alpha.new(:get_question, mock_method)

      # Mock auth layer
      mock_credentials = MockQuestionServiceCredentials_v1alpha.new("get_question")

      Google::Cloud::Dataqna::V1alpha::QuestionService::Stub.stub(:new, mock_stub) do
        Google::Cloud::DataQnA::V1alpha::Credentials.stub(:default, mock_credentials) do
          client = Google::Cloud::DataQnA::V1alpha::Question.new

          # Call method
          response = client.get_question(formatted_name)

          # Verify the response
          assert_equal(expected_response, response)

          # Call method with block
          client.get_question(formatted_name) do |response, operation|
            # Verify the response
            assert_equal(expected_response, response)
            refute_nil(operation)
          end
        end
      end
    end

    it 'invokes get_question with error' do
      # Create request parameters
      formatted_name = Google::Cloud::DataQnA::V1alpha::QuestionServiceClient.question_path("[PROJECT]", "[LOCATION]", "[QUESTION]")

      # Mock Grpc layer
      mock_method = proc do |request|
        assert_instance_of(Google::Cloud::Dataqna::V1alpha::GetQuestionRequest, request)
        assert_equal(formatted_name, request.name)
        raise custom_error
      end
      mock_stub = MockGrpcClientStub_v1alpha.new(:get_question, mock_method)

      # Mock auth layer
      mock_credentials = MockQuestionServiceCredentials_v1alpha.new("get_question")

      Google::Cloud::Dataqna::V1alpha::QuestionService::Stub.stub(:new, mock_stub) do
        Google::Cloud::DataQnA::V1alpha::Credentials.stub(:default, mock_credentials) do
          client = Google::Cloud::DataQnA::V1alpha::Question.new

          # Call method
          err = assert_raises Google::Gax::GaxError, CustomTestError_v1alpha do
            client.get_question(formatted_name)
          end

          # Verify the GaxError wrapped the custom error that was raised.
          assert_match(custom_error.message, err.message)
        end
      end
    end
  end

  describe 'create_question' do
    custom_error = CustomTestError_v1alpha.new "Custom test error for Google::Cloud::DataQnA::V1alpha::QuestionServiceClient#create_question."

    it 'invokes create_question without error' do
      # Create request parameters
      formatted_parent = Google::Cloud::DataQnA::V1alpha::QuestionServiceClient.location_path("[PROJECT]", "[LOCATION]")
      question = {}

      # Create expected grpc response
      name = "name3373707"
      query = "query107944136"
      user_email = "userEmail1921668648"
      expected_response = {
        name: name,
        query: query,
        user_email: user_email
      }
      expected_response = Google::Gax::to_proto(expected_response, Google::Cloud::Dataqna::V1alpha::Question)

      # Mock Grpc layer
      mock_method = proc do |request|
        assert_instance_of(Google::Cloud::Dataqna::V1alpha::CreateQuestionRequest, request)
        assert_equal(formatted_parent, request.parent)
        assert_equal(Google::Gax::to_proto(question, Google::Cloud::Dataqna::V1alpha::Question), request.question)
        OpenStruct.new(execute: expected_response)
      end
      mock_stub = MockGrpcClientStub_v1alpha.new(:create_question, mock_method)

      # Mock auth layer
      mock_credentials = MockQuestionServiceCredentials_v1alpha.new("create_question")

      Google::Cloud::Dataqna::V1alpha::QuestionService::Stub.stub(:new, mock_stub) do
        Google::Cloud::DataQnA::V1alpha::Credentials.stub(:default, mock_credentials) do
          client = Google::Cloud::DataQnA::V1alpha::Question.new

          # Call method
          response = client.create_question(formatted_parent, question)

          # Verify the response
          assert_equal(expected_response, response)

          # Call method with block
          client.create_question(formatted_parent, question) do |response, operation|
            # Verify the response
            assert_equal(expected_response, response)
            refute_nil(operation)
          end
        end
      end
    end

    it 'invokes create_question with error' do
      # Create request parameters
      formatted_parent = Google::Cloud::DataQnA::V1alpha::QuestionServiceClient.location_path("[PROJECT]", "[LOCATION]")
      question = {}

      # Mock Grpc layer
      mock_method = proc do |request|
        assert_instance_of(Google::Cloud::Dataqna::V1alpha::CreateQuestionRequest, request)
        assert_equal(formatted_parent, request.parent)
        assert_equal(Google::Gax::to_proto(question, Google::Cloud::Dataqna::V1alpha::Question), request.question)
        raise custom_error
      end
      mock_stub = MockGrpcClientStub_v1alpha.new(:create_question, mock_method)

      # Mock auth layer
      mock_credentials = MockQuestionServiceCredentials_v1alpha.new("create_question")

      Google::Cloud::Dataqna::V1alpha::QuestionService::Stub.stub(:new, mock_stub) do
        Google::Cloud::DataQnA::V1alpha::Credentials.stub(:default, mock_credentials) do
          client = Google::Cloud::DataQnA::V1alpha::Question.new

          # Call method
          err = assert_raises Google::Gax::GaxError, CustomTestError_v1alpha do
            client.create_question(formatted_parent, question)
          end

          # Verify the GaxError wrapped the custom error that was raised.
          assert_match(custom_error.message, err.message)
        end
      end
    end
  end

  describe 'execute_question' do
    custom_error = CustomTestError_v1alpha.new "Custom test error for Google::Cloud::DataQnA::V1alpha::QuestionServiceClient#execute_question."

    it 'invokes execute_question without error' do
      # Create request parameters
      name = ''
      interpretation_index = 0

      # Create expected grpc response
      name_2 = "name2-1052831874"
      query = "query107944136"
      user_email = "userEmail1921668648"
      expected_response = {
        name: name_2,
        query: query,
        user_email: user_email
      }
      expected_response = Google::Gax::to_proto(expected_response, Google::Cloud::Dataqna::V1alpha::Question)

      # Mock Grpc layer
      mock_method = proc do |request|
        assert_instance_of(Google::Cloud::Dataqna::V1alpha::ExecuteQuestionRequest, request)
        assert_equal(name, request.name)
        assert_equal(interpretation_index, request.interpretation_index)
        OpenStruct.new(execute: expected_response)
      end
      mock_stub = MockGrpcClientStub_v1alpha.new(:execute_question, mock_method)

      # Mock auth layer
      mock_credentials = MockQuestionServiceCredentials_v1alpha.new("execute_question")

      Google::Cloud::Dataqna::V1alpha::QuestionService::Stub.stub(:new, mock_stub) do
        Google::Cloud::DataQnA::V1alpha::Credentials.stub(:default, mock_credentials) do
          client = Google::Cloud::DataQnA::V1alpha::Question.new

          # Call method
          response = client.execute_question(name, interpretation_index)

          # Verify the response
          assert_equal(expected_response, response)

          # Call method with block
          client.execute_question(name, interpretation_index) do |response, operation|
            # Verify the response
            assert_equal(expected_response, response)
            refute_nil(operation)
          end
        end
      end
    end

    it 'invokes execute_question with error' do
      # Create request parameters
      name = ''
      interpretation_index = 0

      # Mock Grpc layer
      mock_method = proc do |request|
        assert_instance_of(Google::Cloud::Dataqna::V1alpha::ExecuteQuestionRequest, request)
        assert_equal(name, request.name)
        assert_equal(interpretation_index, request.interpretation_index)
        raise custom_error
      end
      mock_stub = MockGrpcClientStub_v1alpha.new(:execute_question, mock_method)

      # Mock auth layer
      mock_credentials = MockQuestionServiceCredentials_v1alpha.new("execute_question")

      Google::Cloud::Dataqna::V1alpha::QuestionService::Stub.stub(:new, mock_stub) do
        Google::Cloud::DataQnA::V1alpha::Credentials.stub(:default, mock_credentials) do
          client = Google::Cloud::DataQnA::V1alpha::Question.new

          # Call method
          err = assert_raises Google::Gax::GaxError, CustomTestError_v1alpha do
            client.execute_question(name, interpretation_index)
          end

          # Verify the GaxError wrapped the custom error that was raised.
          assert_match(custom_error.message, err.message)
        end
      end
    end
  end

  describe 'get_user_feedback' do
    custom_error = CustomTestError_v1alpha.new "Custom test error for Google::Cloud::DataQnA::V1alpha::QuestionServiceClient#get_user_feedback."

    it 'invokes get_user_feedback without error' do
      # Create request parameters
      formatted_name = Google::Cloud::DataQnA::V1alpha::QuestionServiceClient.user_feedback_path("[PROJECT]", "[LOCATION]", "[QUESTION]")

      # Create expected grpc response
      name_2 = "name2-1052831874"
      free_form_feedback = "freeFormFeedback713254925"
      expected_response = { name: name_2, free_form_feedback: free_form_feedback }
      expected_response = Google::Gax::to_proto(expected_response, Google::Cloud::Dataqna::V1alpha::UserFeedback)

      # Mock Grpc layer
      mock_method = proc do |request|
        assert_instance_of(Google::Cloud::Dataqna::V1alpha::GetUserFeedbackRequest, request)
        assert_equal(formatted_name, request.name)
        OpenStruct.new(execute: expected_response)
      end
      mock_stub = MockGrpcClientStub_v1alpha.new(:get_user_feedback, mock_method)

      # Mock auth layer
      mock_credentials = MockQuestionServiceCredentials_v1alpha.new("get_user_feedback")

      Google::Cloud::Dataqna::V1alpha::QuestionService::Stub.stub(:new, mock_stub) do
        Google::Cloud::DataQnA::V1alpha::Credentials.stub(:default, mock_credentials) do
          client = Google::Cloud::DataQnA::V1alpha::Question.new

          # Call method
          response = client.get_user_feedback(formatted_name)

          # Verify the response
          assert_equal(expected_response, response)

          # Call method with block
          client.get_user_feedback(formatted_name) do |response, operation|
            # Verify the response
            assert_equal(expected_response, response)
            refute_nil(operation)
          end
        end
      end
    end

    it 'invokes get_user_feedback with error' do
      # Create request parameters
      formatted_name = Google::Cloud::DataQnA::V1alpha::QuestionServiceClient.user_feedback_path("[PROJECT]", "[LOCATION]", "[QUESTION]")

      # Mock Grpc layer
      mock_method = proc do |request|
        assert_instance_of(Google::Cloud::Dataqna::V1alpha::GetUserFeedbackRequest, request)
        assert_equal(formatted_name, request.name)
        raise custom_error
      end
      mock_stub = MockGrpcClientStub_v1alpha.new(:get_user_feedback, mock_method)

      # Mock auth layer
      mock_credentials = MockQuestionServiceCredentials_v1alpha.new("get_user_feedback")

      Google::Cloud::Dataqna::V1alpha::QuestionService::Stub.stub(:new, mock_stub) do
        Google::Cloud::DataQnA::V1alpha::Credentials.stub(:default, mock_credentials) do
          client = Google::Cloud::DataQnA::V1alpha::Question.new

          # Call method
          err = assert_raises Google::Gax::GaxError, CustomTestError_v1alpha do
            client.get_user_feedback(formatted_name)
          end

          # Verify the GaxError wrapped the custom error that was raised.
          assert_match(custom_error.message, err.message)
        end
      end
    end
  end

  describe 'update_user_feedback' do
    custom_error = CustomTestError_v1alpha.new "Custom test error for Google::Cloud::DataQnA::V1alpha::QuestionServiceClient#update_user_feedback."

    it 'invokes update_user_feedback without error' do
      # Create request parameters
      user_feedback = {}

      # Create expected grpc response
      name = "name3373707"
      free_form_feedback = "freeFormFeedback713254925"
      expected_response = { name: name, free_form_feedback: free_form_feedback }
      expected_response = Google::Gax::to_proto(expected_response, Google::Cloud::Dataqna::V1alpha::UserFeedback)

      # Mock Grpc layer
      mock_method = proc do |request|
        assert_instance_of(Google::Cloud::Dataqna::V1alpha::UpdateUserFeedbackRequest, request)
        assert_equal(Google::Gax::to_proto(user_feedback, Google::Cloud::Dataqna::V1alpha::UserFeedback), request.user_feedback)
        OpenStruct.new(execute: expected_response)
      end
      mock_stub = MockGrpcClientStub_v1alpha.new(:update_user_feedback, mock_method)

      # Mock auth layer
      mock_credentials = MockQuestionServiceCredentials_v1alpha.new("update_user_feedback")

      Google::Cloud::Dataqna::V1alpha::QuestionService::Stub.stub(:new, mock_stub) do
        Google::Cloud::DataQnA::V1alpha::Credentials.stub(:default, mock_credentials) do
          client = Google::Cloud::DataQnA::V1alpha::Question.new

          # Call method
          response = client.update_user_feedback(user_feedback)

          # Verify the response
          assert_equal(expected_response, response)

          # Call method with block
          client.update_user_feedback(user_feedback) do |response, operation|
            # Verify the response
            assert_equal(expected_response, response)
            refute_nil(operation)
          end
        end
      end
    end

    it 'invokes update_user_feedback with error' do
      # Create request parameters
      user_feedback = {}

      # Mock Grpc layer
      mock_method = proc do |request|
        assert_instance_of(Google::Cloud::Dataqna::V1alpha::UpdateUserFeedbackRequest, request)
        assert_equal(Google::Gax::to_proto(user_feedback, Google::Cloud::Dataqna::V1alpha::UserFeedback), request.user_feedback)
        raise custom_error
      end
      mock_stub = MockGrpcClientStub_v1alpha.new(:update_user_feedback, mock_method)

      # Mock auth layer
      mock_credentials = MockQuestionServiceCredentials_v1alpha.new("update_user_feedback")

      Google::Cloud::Dataqna::V1alpha::QuestionService::Stub.stub(:new, mock_stub) do
        Google::Cloud::DataQnA::V1alpha::Credentials.stub(:default, mock_credentials) do
          client = Google::Cloud::DataQnA::V1alpha::Question.new

          # Call method
          err = assert_raises Google::Gax::GaxError, CustomTestError_v1alpha do
            client.update_user_feedback(user_feedback)
          end

          # Verify the GaxError wrapped the custom error that was raised.
          assert_match(custom_error.message, err.message)
        end
      end
    end
  end
end

File: google/cloud/dataqna/v1alpha/google-cloud-dataqna-v1alpha-java/gapic-google-cloud-dataqna-v1alpha-java/src/test/java/com/google/cloud/dataqna/v1alpha/QuestionServiceClientTest.java
/*
 * Copyright 2021 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.google.cloud.dataqna.v1alpha;

import com.google.api.gax.core.NoCredentialsProvider;
import com.google.api.gax.grpc.GaxGrpcProperties;
import com.google.api.gax.grpc.testing.LocalChannelProvider;
import com.google.api.gax.grpc.testing.MockGrpcService;
import com.google.api.gax.grpc.testing.MockServiceHelper;
import com.google.api.gax.rpc.ApiClientHeaderProvider;
import com.google.api.gax.rpc.InvalidArgumentException;
import com.google.protobuf.AbstractMessage;
import com.google.protobuf.Any;
import com.google.protobuf.FieldMask;
import com.google.protobuf.Timestamp;
import io.grpc.StatusRuntimeException;
import java.io.IOException;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;
import java.util.UUID;
import javax.annotation.Generated;
import org.junit.After;
import org.junit.AfterClass;
import org.junit.Assert;
import org.junit.Before;
import org.junit.BeforeClass;
import org.junit.Test;

@Generated("by gapic-generator-java")
public class QuestionServiceClientTest {
  private static MockQuestionService mockQuestionService;
  private static MockServiceHelper mockServiceHelper;
  private QuestionServiceClient client;
  private LocalChannelProvider channelProvider;

  @BeforeClass
  public static void startStaticServer() {
    mockQuestionService = new MockQuestionService();
    mockServiceHelper =
        new MockServiceHelper(
            UUID.randomUUID().toString(), Arrays.<MockGrpcService>asList(mockQuestionService));
    mockServiceHelper.start();
  }

  @AfterClass
  public static void stopServer() {
    mockServiceHelper.stop();
  }

  @Before
  public void setUp() throws IOException {
    mockServiceHelper.reset();
    channelProvider = mockServiceHelper.createChannelProvider();
    QuestionServiceSettings settings =
        QuestionServiceSettings.newBuilder()
            .setTransportChannelProvider(channelProvider)
            .setCredentialsProvider(NoCredentialsProvider.create())
            .build();
    client = QuestionServiceClient.create(settings);
  }

  @After
  public void tearDown() throws Exception {
    client.close();
  }

  @Test
  public void getQuestionTest() throws Exception {
    Question expectedResponse =
        Question.newBuilder()
            .setName(QuestionName.of("[PROJECT]", "[LOCATION]", "[QUESTION]").toString())
            .addAllScopes(new ArrayList<String>())
            .setQuery("query107944136")
            .addAllDataSourceAnnotations(new ArrayList<String>())
            .setInterpretError(InterpretError.newBuilder().build())
            .addAllInterpretations(new ArrayList<Interpretation>())
            .setCreateTime(Timestamp.newBuilder().build())
            .setUserEmail("userEmail315299473")
            .setDebugFlags(DebugFlags.newBuilder().build())
            .setDebugInfo(Any.newBuilder().build())
            .build();
    mockQuestionService.addResponse(expectedResponse);

    QuestionName name = QuestionName.of("[PROJECT]", "[LOCATION]", "[QUESTION]");

    Question actualResponse = client.getQuestion(name);
    Assert.assertEquals(expectedResponse, actualResponse);

    List<AbstractMessage> actualRequests = mockQuestionService.getRequests();
    Assert.assertEquals(1, actualRequests.size());
    GetQuestionRequest actualRequest = ((GetQuestionRequest) actualRequests.get(0));

    Assert.assertEquals(name.toString(), actualRequest.getName());
    Assert.assertTrue(
        channelProvider.isHeaderSent(
            ApiClientHeaderProvider.getDefaultApiClientHeaderKey(),
            GaxGrpcProperties.getDefaultApiClientHeaderPattern()));
  }

  @Test
  public void getQuestionExceptionTest() throws Exception {
    StatusRuntimeException exception = new StatusRuntimeException(io.grpc.Status.INVALID_ARGUMENT);
    mockQuestionService.addException(exception);

    try {
      QuestionName name = QuestionName.of("[PROJECT]", "[LOCATION]", "[QUESTION]");
      client.getQuestion(name);
      Assert.fail("No exception raised");
    } catch (InvalidArgumentException e) {
      // Expected exception.
    }
  }

  @Test
  public void getQuestionTest2() throws Exception {
    Question expectedResponse =
        Question.newBuilder()
            .setName(QuestionName.of("[PROJECT]", "[LOCATION]", "[QUESTION]").toString())
            .addAllScopes(new ArrayList<String>())
            .setQuery("query107944136")
            .addAllDataSourceAnnotations(new ArrayList<String>())
            .setInterpretError(InterpretError.newBuilder().build())
            .addAllInterpretations(new ArrayList<Interpretation>())
            .setCreateTime(Timestamp.newBuilder().build())
            .setUserEmail("userEmail315299473")
            .setDebugFlags(DebugFlags.newBuilder().build())
            .setDebugInfo(Any.newBuilder().build())
            .build();
    mockQuestionService.addResponse(expectedResponse);

    String name = "name3373707";

    Question actualResponse = client.getQuestion(name);
    Assert.assertEquals(expectedResponse, actualResponse);

    List<AbstractMessage> actualRequests = mockQuestionService.getRequests();
    Assert.assertEquals(1, actualRequests.size());
    GetQuestionRequest actualRequest = ((GetQuestionRequest) actualRequests.get(0));

    Assert.assertEquals(name, actualRequest.getName());
    Assert.assertTrue(
        channelProvider.isHeaderSent(
            ApiClientHeaderProvider.getDefaultApiClientHeaderKey(),
            GaxGrpcProperties.getDefaultApiClientHeaderPattern()));
  }

  @Test
  public void getQuestionExceptionTest2() throws Exception {
    StatusRuntimeException exception = new StatusRuntimeException(io.grpc.Status.INVALID_ARGUMENT);
    mockQuestionService.addException(exception);

    try {
      String name = "name3373707";
      client.getQuestion(name);
      Assert.fail("No exception raised");
    } catch (InvalidArgumentException e) {
      // Expected exception.
    }
  }

  @Test
  public void createQuestionTest() throws Exception {
    Question expectedResponse =
        Question.newBuilder()
            .setName(QuestionName.of("[PROJECT]", "[LOCATION]", "[QUESTION]").toString())
            .addAllScopes(new ArrayList<String>())
            .setQuery("query107944136")
            .addAllDataSourceAnnotations(new ArrayList<String>())
            .setInterpretError(InterpretError.newBuilder().build())
            .addAllInterpretations(new ArrayList<Interpretation>())
            .setCreateTime(Timestamp.newBuilder().build())
            .setUserEmail("userEmail315299473")
            .setDebugFlags(DebugFlags.newBuilder().build())
            .setDebugInfo(Any.newBuilder().build())
            .build();
    mockQuestionService.addResponse(expectedResponse);

    LocationName parent = LocationName.of("[PROJECT]", "[LOCATION]");
    Question question = Question.newBuilder().build();

    Question actualResponse = client.createQuestion(parent, question);
    Assert.assertEquals(expectedResponse, actualResponse);

    List<AbstractMessage> actualRequests = mockQuestionService.getRequests();
    Assert.assertEquals(1, actualRequests.size());
    CreateQuestionRequest actualRequest = ((CreateQuestionRequest) actualRequests.get(0));

    Assert.assertEquals(parent.toString(), actualRequest.getParent());
    Assert.assertEquals(question, actualRequest.getQuestion());
    Assert.assertTrue(
        channelProvider.isHeaderSent(
            ApiClientHeaderProvider.getDefaultApiClientHeaderKey(),
            GaxGrpcProperties.getDefaultApiClientHeaderPattern()));
  }

  @Test
  public void createQuestionExceptionTest() throws Exception {
    StatusRuntimeException exception = new StatusRuntimeException(io.grpc.Status.INVALID_ARGUMENT);
    mockQuestionService.addException(exception);

    try {
      LocationName parent = LocationName.of("[PROJECT]", "[LOCATION]");
      Question question = Question.newBuilder().build();
      client.createQuestion(parent, question);
      Assert.fail("No exception raised");
    } catch (InvalidArgumentException e) {
      // Expected exception.
    }
  }

  @Test
  public void createQuestionTest2() throws Exception {
    Question expectedResponse =
        Question.newBuilder()
            .setName(QuestionName.of("[PROJECT]", "[LOCATION]", "[QUESTION]").toString())
            .addAllScopes(new ArrayList<String>())
            .setQuery("query107944136")
            .addAllDataSourceAnnotations(new ArrayList<String>())
            .setInterpretError(InterpretError.newBuilder().build())
            .addAllInterpretations(new ArrayList<Interpretation>())
            .setCreateTime(Timestamp.newBuilder().build())
            .setUserEmail("userEmail315299473")
            .setDebugFlags(DebugFlags.newBuilder().build())
            .setDebugInfo(Any.newBuilder().build())
            .build();
    mockQuestionService.addResponse(expectedResponse);

    String parent = "parent-995424086";
    Question question = Question.newBuilder().build();

    Question actualResponse = client.createQuestion(parent, question);
    Assert.assertEquals(expectedResponse, actualResponse);

    List<AbstractMessage> actualRequests = mockQuestionService.getRequests();
    Assert.assertEquals(1, actualRequests.size());
    CreateQuestionRequest actualRequest = ((CreateQuestionRequest) actualRequests.get(0));

    Assert.assertEquals(parent, actualRequest.getParent());
    Assert.assertEquals(question, actualRequest.getQuestion());
    Assert.assertTrue(
        channelProvider.isHeaderSent(
            ApiClientHeaderProvider.getDefaultApiClientHeaderKey(),
            GaxGrpcProperties.getDefaultApiClientHeaderPattern()));
  }

  @Test
  public void createQuestionExceptionTest2() throws Exception {
    StatusRuntimeException exception = new StatusRuntimeException(io.grpc.Status.INVALID_ARGUMENT);
    mockQuestionService.addException(exception);

    try {
      String parent = "parent-995424086";
      Question question = Question.newBuilder().build();
      client.createQuestion(parent, question);
      Assert.fail("No exception raised");
    } catch (InvalidArgumentException e) {
      // Expected exception.
    }
  }

  @Test
  public void executeQuestionTest() throws Exception {
    Question expectedResponse =
        Question.newBuilder()
            .setName(QuestionName.of("[PROJECT]", "[LOCATION]", "[QUESTION]").toString())
            .addAllScopes(new ArrayList<String>())
            .setQuery("query107944136")
            .addAllDataSourceAnnotations(new ArrayList<String>())
            .setInterpretError(InterpretError.newBuilder().build())
            .addAllInterpretations(new ArrayList<Interpretation>())
            .setCreateTime(Timestamp.newBuilder().build())
            .setUserEmail("userEmail315299473")
            .setDebugFlags(DebugFlags.newBuilder().build())
            .setDebugInfo(Any.newBuilder().build())
            .build();
    mockQuestionService.addResponse(expectedResponse);

    String name = "name3373707";
    int interpretationIndex = -1924823077;

    Question actualResponse = client.executeQuestion(name, interpretationIndex);
    Assert.assertEquals(expectedResponse, actualResponse);

    List<AbstractMessage> actualRequests = mockQuestionService.getRequests();
    Assert.assertEquals(1, actualRequests.size());
    ExecuteQuestionRequest actualRequest = ((ExecuteQuestionRequest) actualRequests.get(0));

    Assert.assertEquals(name, actualRequest.getName());
    Assert.assertEquals(interpretationIndex, actualRequest.getInterpretationIndex());
    Assert.assertTrue(
        channelProvider.isHeaderSent(
            ApiClientHeaderProvider.getDefaultApiClientHeaderKey(),
            GaxGrpcProperties.getDefaultApiClientHeaderPattern()));
  }

  @Test
  public void executeQuestionExceptionTest() throws Exception {
    StatusRuntimeException exception = new StatusRuntimeException(io.grpc.Status.INVALID_ARGUMENT);
    mockQuestionService.addException(exception);

    try {
      String name = "name3373707";
      int interpretationIndex = -1924823077;
      client.executeQuestion(name, interpretationIndex);
      Assert.fail("No exception raised");
    } catch (InvalidArgumentException e) {
      // Expected exception.
    }
  }

  @Test
  public void getUserFeedbackTest() throws Exception {
    UserFeedback expectedResponse =
        UserFeedback.newBuilder()
            .setName(UserFeedbackName.of("[PROJECT]", "[LOCATION]", "[QUESTION]").toString())
            .setFreeFormFeedback("freeFormFeedback-1509261067")
            .build();
    mockQuestionService.addResponse(expectedResponse);

    UserFeedbackName name = UserFeedbackName.of("[PROJECT]", "[LOCATION]", "[QUESTION]");

    UserFeedback actualResponse = client.getUserFeedback(name);
    Assert.assertEquals(expectedResponse, actualResponse);

    List<AbstractMessage> actualRequests = mockQuestionService.getRequests();
    Assert.assertEquals(1, actualRequests.size());
    GetUserFeedbackRequest actualRequest = ((GetUserFeedbackRequest) actualRequests.get(0));

    Assert.assertEquals(name.toString(), actualRequest.getName());
    Assert.assertTrue(
        channelProvider.isHeaderSent(
            ApiClientHeaderProvider.getDefaultApiClientHeaderKey(),
            GaxGrpcProperties.getDefaultApiClientHeaderPattern()));
  }

  @Test
  public void getUserFeedbackExceptionTest() throws Exception {
    StatusRuntimeException exception = new StatusRuntimeException(io.grpc.Status.INVALID_ARGUMENT);
    mockQuestionService.addException(exception);

    try {
      UserFeedbackName name = UserFeedbackName.of("[PROJECT]", "[LOCATION]", "[QUESTION]");
      client.getUserFeedback(name);
      Assert.fail("No exception raised");
    } catch (InvalidArgumentException e) {
      // Expected exception.
    }
  }

  @Test
  public void getUserFeedbackTest2() throws Exception {
    UserFeedback expectedResponse =
        UserFeedback.newBuilder()
            .setName(UserFeedbackName.of("[PROJECT]", "[LOCATION]", "[QUESTION]").toString())
            .setFreeFormFeedback("freeFormFeedback-1509261067")
            .build();
    mockQuestionService.addResponse(expectedResponse);

    String name = "name3373707";

    UserFeedback actualResponse = client.getUserFeedback(name);
    Assert.assertEquals(expectedResponse, actualResponse);

    List<AbstractMessage> actualRequests = mockQuestionService.getRequests();
    Assert.assertEquals(1, actualRequests.size());
    GetUserFeedbackRequest actualRequest = ((GetUserFeedbackRequest) actualRequests.get(0));

    Assert.assertEquals(name, actualRequest.getName());
    Assert.assertTrue(
        channelProvider.isHeaderSent(
            ApiClientHeaderProvider.getDefaultApiClientHeaderKey(),
            GaxGrpcProperties.getDefaultApiClientHeaderPattern()));
  }

  @Test
  public void getUserFeedbackExceptionTest2() throws Exception {
    StatusRuntimeException exception = new StatusRuntimeException(io.grpc.Status.INVALID_ARGUMENT);
    mockQuestionService.addException(exception);

    try {
      String name = "name3373707";
      client.getUserFeedback(name);
      Assert.fail("No exception raised");
    } catch (InvalidArgumentException e) {
      // Expected exception.
    }
  }

  @Test
  public void updateUserFeedbackTest() throws Exception {
    UserFeedback expectedResponse =
        UserFeedback.newBuilder()
            .setName(UserFeedbackName.of("[PROJECT]", "[LOCATION]", "[QUESTION]").toString())
            .setFreeFormFeedback("freeFormFeedback-1509261067")
            .build();
    mockQuestionService.addResponse(expectedResponse);

    UserFeedback userFeedback = UserFeedback.newBuilder().build();
    FieldMask updateMask = FieldMask.newBuilder().build();

    UserFeedback actualResponse = client.updateUserFeedback(userFeedback, updateMask);
    Assert.assertEquals(expectedResponse, actualResponse);

    List<AbstractMessage> actualRequests = mockQuestionService.getRequests();
    Assert.assertEquals(1, actualRequests.size());
    UpdateUserFeedbackRequest actualRequest = ((UpdateUserFeedbackRequest) actualRequests.get(0));

    Assert.assertEquals(userFeedback, actualRequest.getUserFeedback());
    Assert.assertEquals(updateMask, actualRequest.getUpdateMask());
    Assert.assertTrue(
        channelProvider.isHeaderSent(
            ApiClientHeaderProvider.getDefaultApiClientHeaderKey(),
            GaxGrpcProperties.getDefaultApiClientHeaderPattern()));
  }

  @Test
  public void updateUserFeedbackExceptionTest() throws Exception {
    StatusRuntimeException exception = new StatusRuntimeException(io.grpc.Status.INVALID_ARGUMENT);
    mockQuestionService.addException(exception);

    try {
      UserFeedback userFeedback = UserFeedback.newBuilder().build();
      FieldMask updateMask = FieldMask.newBuilder().build();
      client.updateUserFeedback(userFeedback, updateMask);
      Assert.fail("No exception raised");
    } catch (InvalidArgumentException e) {
      // Expected exception.
    }
  }
}
